Add network controlled car

RileyEv opened the pull request:
Car now syncs position and rotation across the network.

Resolves #17 
Resolves #22 
Resolves #23 
Resolves #33 
Resolves #35

RileyEv committed:
Spawn car when client starts

RileyEv committed:
Uncomplete added 'random' spawn locations

RileyEv committed:
Removes Spawner

RileyEv committed:
WIP: script to copy and symlink

RileyEv committed:
All changes

RileyEv committed:
Merge branch 'editor-tools' into add-car-spawn

RileyEv committed:
Add spawning and location synchronisation

RileyEv committed:
Added rotation sync

RileyEv committed:
Fix Camera following

ChrisGora commented:
I will go through the code and try to do a small refactor. Rn the words "Fuck it" were said many times. 😂

liamdalg commented on ProjectFiles/Assets/Scripts/Network/Server.cs:
```diff
@@ -87,35 +94,78 @@ public void HandleNetworkEvents()
                             var readerContext = default(DataStreamReader.Context);
                             var ev = (ClientNetworkEvent) reader.ReadByte(ref readerContext);
-                            HandleEvent(connections[i], endpoint, ev, reader, readerContext);
+                            HandleEvent(connections[i], endpoint, ev, reader, readerContext, i);
```
I think this won't always work? I can't currently test it, but the `i` variable may not always refer to the same connection since the connections list may change. Instead I think we should use the internal connection ID.

```suggestion
                            HandleEvent(connections[i], endpoint, ev, reader, readerContext, connections[i].InternalId);
```

liamdalg commented on ProjectFiles/Assets/Scripts/Network/Server.cs:
```diff
@@ -87,35 +94,78 @@ public void HandleNetworkEvents()
                             Debug.Log($"Server: {endpoint.IpAddress()}:{endpoint.Port} disconnected.");
                             connections[i] = default(NetworkConnection);
+
+                            var playerID = connectionPlayerIDs[i];
```
Same thing as above.

```suggestion
                            var playerID = connectionPlayerIDs[connections[i].InternalId];
```

liamdalg commented on ProjectFiles/Assets/Scripts/Network/Server.cs:
```diff
@@ -87,35 +94,78 @@ public void HandleNetworkEvents()
+
+                            var playerID = connectionPlayerIDs[i];
+                            world.DestroyPlayer(playerID);
+                            connectionPlayerIDs.Remove(i);
```
Again...

```suggestion
                            connectionPlayerIDs.Remove(connections[i].InternalId);
```

liamdalg commented on ProjectFiles/Assets/Scripts/Network/Server.cs:
```diff
@@ -87,35 +94,78 @@ public void HandleNetworkEvents()
                         }
                         case NetworkEvent.Type.Disconnect:
                             Debug.Log($"Server: {endpoint.IpAddress()}:{endpoint.Port} disconnected.");
                             connections[i] = default(NetworkConnection);
```
If the InternalId change is correct, you'll need to move this line `connections[i] = default(NetworkConnection);` to after we remove the ID from the dict, i.e. at the end of the case statement.

liamdalg reviewed (changes requested):
Seems good overall, really good job (sorry I wasn't there!!) Just check my changes wrt. the InternalId stuff, I could be wrong.

ChrisGora commented:
Will have a closer look. Overall we tried to be careful with different IDs and indexes, but maybe we missed that one 🙈

RileyEv commented:
Yes we do need to change that. We knew that the index would change because we had big discussion about whether we should use a list or dictionary, but then clearly forgot when indexing the dictionary 🙈

RileyEv commented:
Also before we merge we need to make this work without a server working.

ghost commented:
> Also before we merge we need to make this work without a server working.

In the option thing that lets you choose between client and server maybe have a third option called standalone or something that will spawn the car without a server?

ghost commented:
Who wants to implement Liam's change? Or click the buttons or whatever needs doing...

RileyEv commented:
Yeah that’s what I was thinking

ghost commented:
Happy to do some of this at the weekend.

liamdalg commented on extraeditor.sh:
```diff
@@ -0,0 +1,18 @@
+rm -rf $newfolder/ProjectFiles/Assets $newfolder/ProjectFiles/ProjectSettings
+
+ln -s $currdirectory/ProjectFiles/Assets $newfolder/ProjectFiles/Assets
+ln -s $currdirectory/ProjectFiles/ProjectSettings $newfolder/ProjectFiles/ProjectSetings
```
Typo?

```suggestion
ln -s $currdirectory/ProjectFiles/ProjectSettings $newfolder/ProjectFiles/ProjectSettings
```

RileyEv replied:
Oh my fucking god that’s why it wasn’t working 😤

liamdalg committed:
load server config from file

liamdalg committed:
load server cfg in constructor to prevent memory leak with NativeList

liamdalg committed:
catch FileNotFoundException when loading config

liamdalg committed:
move config load/save to ServerConfig

liamdalg committed:
forgot default config path

Commit:
Pretty print default config file

Commit:
Ignore config file

Commit:
Make DefaultConfig private + readonly

Commit:
Merge pull request #36 from BigMacintosh/server-config

Server config

Commit:
Remove client ID in location update, use internal connection ids

Commit:
Fix symlink paths

Co-Authored-By: Liam Dalgarno <32525406+liamdalg@users.noreply.github.com>

RileyEv committed:
Change how writers are created

Writer is now returned with identification byte already added

RileyEv committed:
Merge remote-tracking branch 'origin/add-car-spawn' into add-car-spawn

RileyEv committed:
Added Standalone version

- Now defaults to standalone.
- Introduces an IClient interface for Clients
- Creates DummyClient, which spawns a car and ignores all network requests.

RileyEv commented:
This now also resolves #17

RileyEv commented:
We also need to fix the HUD so it works like the Camera. Currently its attached to the old car prefab and not the new car that is spawned.

RileyEv committed:
Merge pull request #38 from BigMacintosh/add-standalone-mode

Add Standalone version

RileyEv committed:
Bug Fix: Cannot set car controllable before storing it

Commit:
Merge pull request #37 from BigMacintosh/refactor-datastreamwriter

Change how DataStreamWriters are created

ChrisGora commented on ProjectFiles/Assets/Scripts/Game/ServerGameLoop.cs:
```diff
@@ -6,18 +6,18 @@ public class ServerGameLoop : IGameLoop
             
             // Create world
-            world = new World(spawner);
+            world = new World();
```
What are we using now instead of the spawner object?

ChrisGora reviewed (approved):
Love the changes <3

RileyEv committed:
Just nothing
